gui/webdisplay/src/TWebWindowsManager.cxx

```diff
@@ -396,26 +400,48 @@ void ROOT::Experimental::TWebWindowsManager::TestProg(TString &prog, const std::
 ///               $height - widget height
 ///
 ///  If allowed, same window can be displayed several times (like for TCanvas)
-
-bool ROOT::Experimental::TWebWindowsManager::Show(ROOT::Experimental::TWebWindow &win, const std::string &_where)
+///  Following parameters can be configured in rootrc file:
+///
+///   WebGui.Chrome:  full path to Google Chrome executable
+///   WebGui.ChromeBatch: command to start chrome in batch
+///   WebGui.ChromeInteractive: command to start chrome in interactive mode
+///   WebGui.Firefox: full path to Mozialla Firefox executable
+///   WebGui.FirefoxBatch: command to start Firefox in batch mode
+///   WebGui.FirefoxInteractive: command to start Firefox in interactive mode
+///   WebGui.FirefoxProfile: name of Firefox profile to use
+///   WebGui.FirefoxProfilePath: file path to Firefox profile
+///   WebGui.FirefoxRandomProfile: usage of random Firefox profile -1 never, 0 - only for batch mode (dflt), 1 - always
+///   WebGui.LaunchTmout: time required to start process in seconds (default 30 s)
+///   WebGui.OperationTmout: time required to perform WebWindow operation like execute command or update drawings
+
+unsigned ROOT::Experimental::TWebWindowsManager::Show(ROOT::Experimental::TWebWindow &win, bool batch_mode, const std::string &_where)
 {
+   // we book manager mutex for a longer operation,
+   // TWebWindowManagerGuard grd(*this);
+   std::lock_guard<std::recursive_mutex> grd(fMutex);
+
    if (!fServer) {
       R__ERROR_HERE("WebDisplay") << "Server instance not exists to show window";
-      return false;
+      return 0;
+   }
+
+   if (!batch_mode && gROOT->IsWebDisplayBatch()) {
+      R__ERROR_HERE("WebDisplay") << "Cannot show normal web window in batch mode";
+      return 0;
    }
 
-   std::string key;
-   int ntry = 1000;
+   std::string key, rmdir;
+   int ntry = 100000;
 
    do {
       key = std::to_string(gRandom->Integer(0x100000));
    } while ((--ntry > 0) && win.HasKey(key));
    if (ntry == 0) {
       R__ERROR_HERE("WebDisplay") << "Fail to create unique key for the window";
-      return false;
+      return 0;
    }
 
-   std::string addr = GetUrl(win, false);
+   std::string addr = GetUrl(win, batch_mode, false);
    if (addr.find("?") != std::string::npos)
       addr.append("&key=");
    else
```

Comment:
Comment on lines 455++: please please please split this into separate lines! Else this totally looks like a complex boolean expression, not four simple ones!

Reply:
Comment on lines 455++: these seem to be "exclusive" booleans. Please introduce an `enum EWhere { kNative, kLocal, kChrome, kFirefox}`.